src/ray/raylet/local_object_manager.cc

```diff
@@ -274,39 +278,65 @@ void LocalObjectManager::AddSpilledUrls(
-          if (!url_ref_count_.contains(base_url_it->second)) {
-            url_ref_count_[base_url_it->second] = 1;
-          } else {
-            url_ref_count_[base_url_it->second] += 1;
-          }
-          spilled_objects_url_.emplace(object_id, object_url);
+    const auto unpin_callback = [this, object_id, object_url, callback,
+                                 num_remaining](Status status) {
+      RAY_CHECK_OK(status);
+      // Unpin the object.
+      auto it = objects_pending_spill_.find(object_id);
+      RAY_CHECK(it != objects_pending_spill_.end());
+      num_bytes_pending_spill_ -= it->second->GetSize();
+      objects_pending_spill_.erase(it);
+
+      // Update the object_id -> url_ref_count to use it for deletion later.
+      // We need to track the references here because a single file can contain
+      // multiple objects, and we shouldn't delete the file until
+      // all the objects are gone out of scope.
+      // object_url is equivalent to url_with_offset.
+      auto parsed_url = ParseURL(object_url);
+      const auto base_url_it = parsed_url->find("url");
+      RAY_CHECK(base_url_it != parsed_url->end());
+      if (!url_ref_count_.contains(base_url_it->second)) {
+        url_ref_count_[base_url_it->second] = 1;
+      } else {
+        url_ref_count_[base_url_it->second] += 1;
+      }
+      spilled_objects_url_.emplace(object_id, object_url);
 
-          (*num_remaining)--;
-          if (*num_remaining == 0 && callback) {
-            callback(status);
-          }
-        }));
+      (*num_remaining)--;
+      if (*num_remaining == 0 && callback) {
+        callback(status);
+      }
+    };
+
+    if (RayConfig::instance().ownership_based_object_directory_enabled()) {
+      // TODO(Clark): Don't send RPC to owner if we're fulfilling an owner-initiated
+      // spill RPC.
+      rpc::AddSpilledUrlRequest request;
+      request.set_object_id(object_id.Binary());
+      request.set_spilled_url(object_url);
+      request.set_spilled_node_id(node_id_object_spilled.Binary());
+      request.set_size(it->second->GetSize());
+
+      auto it = pinned_object_owners_.find(object_id);
+      RAY_CHECK(it != pinned_object_owners_.end());
+      auto owner_client = owner_client_pool_.GetOrConnect(it->second);
+      // Send spilled URL, spilled node ID, and object size to owner.
+      owner_client->AddSpilledUrl(
+          request, [unpin_callback, object_id](Status status,
+                                               const rpc::AddSpilledUrlReply &reply) {
+            if (!status.ok()) {
+              RAY_LOG(ERROR) << "Failed to send spilled URL to owner for " << object_id;
+            }
+            unpin_callback(status);
+          });
+    } else {
+      // Write to object directory. Wait for the write to finish before
+      // releasing the object to make sure that the spilled object can
+      // be retrieved by other raylets.
+      RAY_CHECK_OK(object_info_accessor_.AsyncAddSpilledUrl(
```

Comment:
Not related to this PR, but this code makes me wonder if we should be implementing the OBOD behind the object_info_accessor_ interface. That would make it easier to switch the feature on/off eventually.

Reply:
I briefly looked at that approach, thinking that it would be nice if we could have a single `ObjectDirectory` implementation that took an implementer of `ObjectInfoAccessor`, either GCS-backed or owner-core-worker-backed. There were a few reasons I decided to not go that route (for now):

1. The simplest short-term ownership-based object location subscription implements a different protocol: logically versioned snapshots of the location state, instead of location state update events. I couldn't immediately see a good interface for `ObjectInfoAccessor::AsyncSubscribeToLocations` that could support both protocols. This is something that I'd like to reexamine once we support gRPC server-side streaming in Ray.
2. The `ObjectInfoAccessor` is currently a subinterface for the GCS client, and decoupling _just_ that accessor from the GCS client seemed like it could be messy. The other option would be pulling out all of the accessor subinterfaces into a sibling `accessor` module (or the like), allowing any easy path for those interfaces to be implemented by core workers and/or raylets as more of the GCS state becomes decentralized.
3. I thought that there should be a broader discussion around refactoring the core worker client and raylet client interfaces to adopt something closer to the GCS client model, consisting of application-level subinterfaces that are mostly decoupled from the gRPC layer and whose implementers do the application-level <--> transport level conversion and make the underlying RPC calls, making the users of those subinterfaces mostly transport layer agnostic. I personally like that model, but it'd be a large refactor.
4. It was my impression that the general design of the OBOD was agreed upon, and I didn't want to come in towards the end of the implementation and rewrite all of it. 😄 

(1) and (3) were my biggest reasons for not running with that route. Do you think this is something that we should reexamine once server-side streaming is supported in Ray, or do you think this is something that we should figure out sooner (not in this PR obviously, but in the short-term)?

Reply:
Ah I see. Thanks for investigating that! Very thorough reasoning. I'm okay with not refactoring, since hopefully the OBOD will be on by default soon and it shouldn't be too hard to remove the old code.